machineconfiguration/v1/types.go

```diff
@@ -0,0 +1,389 @@
+
+	// reason is the reason for the condition's last transition.  Reasons are CamelCase
+	Reason string `json:"reason,omitempty"`
+
+	// message provides additional information about the current condition.
+	// This is only to be consumed by humans.
+	Message string `json:"message,omitempty"`
+}
+
+// ControllerConfigStatusConditionType valid conditions of a machineconfigpool
+type ControllerConfigStatusConditionType string
+
+const (
+	TemplateContollerRunning   ControllerConfigStatusConditionType = "TemplateContollerRunning"
+	TemplateContollerCompleted ControllerConfigStatusConditionType = "TemplateContollerCompleted"
+	TemplateContollerFailing   ControllerConfigStatusConditionType = "TemplateContollerFailing"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ControllerConfigList is a list of ControllerConfig resources
+type ControllerConfigList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []ControllerConfig `json:"items"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineConfigPool describes a pool of MachineConfigs.
+type MachineConfigPool struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   MachineConfigPoolSpec   `json:"spec"`
+	Status MachineConfigPoolStatus `json:"status"`
+}
+
+// MachineConfigPoolSpec is the spec for MachineConfigPool resource.
+type MachineConfigPoolSpec struct {
+	// Label selector for MachineConfigs.
+	// Refer https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/ on how label and selectors work.
+	MachineConfigSelector *metav1.LabelSelector `json:"machineConfigSelector,omitempty"`
+
+	// Node selector for Machines
+	NodeSelector *metav1.LabelSelector `json:"nodeSelector,omitempty"`
+
+	// If true, changes to this machine config pool should be stopped.
+	// This includes generating new desiredMachineConfig and update of machines.
+	Paused bool `json:"paused"`
+
+	// MaxUnavailable specifies the percentage or constant number of machines that can be updating at any given time.
+	// default is 1.
+	MaxUnavailable *intstr.IntOrString `json:"maxUnavailable"`
+}
+
+// MachineConfigPoolStatus is the status for MachineConfigPool resource.
+type MachineConfigPoolStatus struct {
+	// The generation observed by the controller.
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// The current MachineConfig object for the machine config pool.
+	Configuration MachineConfigPoolStatusConfiguration `json:"configuration"`
+
+	// Total number of machines in the machine config pool.
+	MachineCount int32 `json:"machineCount"`
+
+	// Total number of machines targeted by the pool that have the CurrentMachineConfig as their config.
+	UpdatedMachineCount int32 `json:"updatedMachineCount"`
+
+	// Total number of ready machines targeted by the pool.
```

Comment:
What is "ready"?  Kubelets that are reporting [the `Ready` condition][1]?  Do we handle machines that aren't running kubelets?  Maybe they're performing some other service directly, and aren't functioning as Kubernetes nodes.

[1]: https://kubernetes.io/docs/concepts/architecture/nodes/#condition

Reply:
ready is described in code (we have a func for it which also takes into account kubelet readiness). And no, we only deal with Kubelet-managed machines today.